source/common/io/io_uring_impl.cc

```diff
@@ -34,47 +34,76 @@ void IoUringFactoryImpl::onServerInitialized() {
 }
 
 IoUringImpl::IoUringImpl(uint32_t io_uring_size, bool use_submission_queue_polling)
-    : io_uring_size_(io_uring_size), cqes_(io_uring_size_, nullptr) {
+    : cqes_(io_uring_size, nullptr) {
   struct io_uring_params p {};
   if (use_submission_queue_polling) {
     p.flags |= IORING_SETUP_SQPOLL;
   }
-  int ret = io_uring_queue_init_params(io_uring_size_, &ring_, &p);
+  // TODO (soulxu): According to the man page: `By default, the CQ ring will have twice the number
```

Comment:
Is it possible that the queue can be starved and then prevent Envoy from making progress. For example, imagine the queue size of 10 and Envoy worker has 1 listening socket and 9 active TCP sockets. The queue may have 10 operations, 1 for listening socket and 9 for reading operations. Next Envoy wants to write into one of the sockets, it wants to place write operation but the queue is full. Will this cause problems?

Reply:
For the SQ ring, we did see the ring can be full when there are too many iouring requests. Then the submission of new request will be rejected. Although we have configure option for the SQ ring size, that doesn't make sense for production use, since the user can't assume how much load the envoy will get. In the end, I'm thinking of some way to enable the envoy to wait until there is more room in the ring, then continue to submit the requests. I leave that as a todo https://github.com/envoyproxy/envoy/issues/28395, I want to improve that after we have iouring implementation first.

For the CQ ring, as the doc said, it will have twice the number of entries than the SQ ring default. At least for now, I don't have any idea for this now, I didn't see a case CQ ring is full in our tests. I also don't know in which case we will get twice the number of request completion that the request we submitted. But yes, we should figure that out, so I leave a todo here.

Reply:
Ok, I understand. Reading a bit more on this, this just limits how many requests can be submitted in one go, not the overall limit of kernel queue. It is a bit confusing with multiple different queue size values in io_uring_queue_init and io_uring_setup. 

It looks to me that what this code should never overflow the max size of the submission batch, since we never process more than io_uring_size_ of operations one wake up. If some test fails, maybe the value passed to the io_uring_queue_init_params just needs some overhead added. Not sure what it needs to be though.

Reply:
> It looks to me that what this code should never overflow the max size of the submission batch, since we never process more than io_uring_size_ of operations one wake up. If some test fails, maybe the value passed to the io_uring_queue_init_params just needs some overhead added. Not sure what it needs to be though.

As I remember some tests failed flaky due to the uring size. Maybe I should add some stat to see how many requests in one wake up, then we will know what happened. Thanks for the reminder, I will keep this in mind, and figure out in the future.